lib/active_model/serializer/type.rb

```diff
@@ -16,8 +16,24 @@ module ClassMethods
         # @example
         #   class AdminAuthorSerializer < ActiveModel::Serializer
         #     type 'authors'
-        def type(type)
-          self._type = type && type.to_s
+        #
+        # @example
+        #   class AdminAuthorSerializer < ActiveModel::Serializer
+        #     type { |object| object.class.name }
+        #
+        # @example
+        #   class AdminAuthorSerializer < ActiveModel::Serializer
+        #     type proc { |object| object.class.name }
+        #
+        def type(type = nil, &block)
+          type = block if block_given?
+          return unless type
+
+          self._type = if type.respond_to?(:call) && type.arity == 1
+                         type
+                       else
+                         type.to_s
```

Comment:
The `.to_s` part doesn't really belong here I think (which would allow to simplify this method).